src/widget/widget.cpp

```diff
@@ -931,34 +931,60 @@ void Widget::reloadHistory()
     }
 }
 
+/**
+ * @brief Plays a sound via the audioNotification AudioSink
+ * @param sound Sound to play
+ * @param loop if true, loop the sound until onStopNotification() is called
+ */
+void Widget::playNotificationSound(IAudioSink::Sound sound, bool loop) {
+    if(audioNotification == nullptr) {
+        audioNotification = std::unique_ptr<IAudioSink>(Audio::getInstance().makeSink());
+        if(audioNotification == nullptr) {
+            qDebug() << "Failed to allocate AudioSink";
+            return;
+        }
+    }
+
+    connect(audioNotification.get(), &IAudioSink::finishedPlaying,
+            this, &Widget::cleanupNotificationSound);
+
+    if(loop) {
+        audioNotification->startLoop();
+    }
+
+    audioNotification->playMono16Sound(sound);
+}
+
+void Widget::cleanupNotificationSound()
+{
+    audioNotification.reset();
+}
+
 void Widget::incomingNotification(uint32_t friendId)
 {
     newFriendMessageAlert(friendId, false);
 
-    Audio& audio = Audio::getInstance();
-    audio.startLoop();
-    audio.playMono16Sound(Audio::getSound(Audio::Sound::IncomingCall));
+    // loop until call answered or rejected
+    playNotificationSound(IAudioSink::Sound::IncomingCall, true);
 }
 
 void Widget::outgoingNotification()
 {
-    Audio& audio = Audio::getInstance();
-    audio.startLoop();
-    audio.playMono16Sound(Audio::getSound(Audio::Sound::OutgoingCall));
+    // loop until call answered or rejected
+    playNotificationSound(IAudioSink::Sound::OutgoingCall, true);
 }
 
 void Widget::onCallEnd()
```

Comment:
Offtop. But why **Widget** react on call end and play end sound? Who, if not him?)